File: tests/fixtures/sample.cs
// Sample.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Demo
{
    public enum Status
    {
        Ok,
        Error,
        Unknown
    }

    public struct Point
    {
        public double X;
        public double Y;

        public double Distance() => Math.Sqrt(X * X + Y * Y);
    }

    public interface ICalculator
    {
        int Add(int a, int b);
    }

    public record Person(string Name, int Age);

    public class Calculator : ICalculator
    {
        private readonly int _base;

        public Calculator(int baseValue)
        {
            _base = baseValue;
        }

        public int Add(int a, int b)
        {
            return a + b + _base;
        }

        public T Multiply<T>(T a, T b) where T : struct
        {
            dynamic x = a;
            dynamic y = b;
            return (T)(x * y);
        }

        public async Task<int> ComputeAsync(IEnumerable<int> values)
        {
            await Task.Delay(10);

            int result = values
                .Where(v => v > 0)
                .Select(v => v * 2)
                .Sum();

            return result;
        }
    }

    public static class Program
    {
        public static async Task Main(string[] args)
        {
            var calc = new Calculator(10);

            List<int> numbers = new() { 1, 2, 3, 4 };

            int sum = await calc.ComputeAsync(numbers);

            int result = calc.Add(sum, 5);

            Point p = new() { X = 3.0, Y = 4.0 };

            string sizeCategory = result switch
            {
                > 20 => "Large",
                0 => "Zero",
                _ => "Small"
            };

            try
            {
                if (result > 10)
                {
                    Console.WriteLine("Result is greater than 10");
                }
                else
                {
                    Console.WriteLine("Result is 10 or less");
                }

                for (int i = 0; i < 3; i++)
                {
                    Console.WriteLine(i);
                }

                int counter = 0;
                while (counter < 2)
                {
                    Console.WriteLine(counter++);
                }

                object obj = p;

                if (obj is Point pt && pt.X > 0)
                {
                    Console.WriteLine($"Point X: {pt.X}");
                }

                void LocalFunction()
                {
                    Console.WriteLine("Inside local function");
                }

                LocalFunction();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            finally
            {
                Console.WriteLine("Done.");
            }

            var person = new Person("Alice", 30);
            Console.WriteLine($"{person.Name} is {person.Age}");
        }
    }
}

File: tests/fixtures/sample.lua
-- sample.lua

-- Global variable
GLOBAL_VALUE = 42

-- Local variable
local pi = 3.14159

-- Table (object-like)
local Point = {}
Point.__index = Point

function Point:new(x, y)
    local obj = setmetatable({}, self)
    obj.x = x
    obj.y = y
    return obj
end

function Point:distance()
    return math.sqrt(self.x * self.x + self.y * self.y)
end

-- Enum-like table
local Status = {
    OK = 1,
    ERROR = 2,
    UNKNOWN = 3
}

-- Generic function with varargs
local function sum(...)
    local total = 0
    for _, value in ipairs({ ... }) do
        total = total + value
    end
    return total
end

-- Higher-order function
local function apply(fn, value)
    return fn(value)
end

-- Closure
local function make_counter(start)
    local count = start or 0
    return function()
        count = count + 1
        return count
    end
end

-- Coroutine example
local function coroutine_example()
    return coroutine.create(function()
        for i = 1, 3 do
            coroutine.yield(i)
        end
    end)
end

-- Main execution block
local function main()
    local numbers = { 1, 2, 3, 4 }
    local result = sum(table.unpack(numbers))

    local p = Point:new(3, 4)

    if result > 5 then
        print("Result is large")
    elseif result == 0 then
        print("Result is zero")
    else
        print("Result is small")
    end

    -- Numeric for
    for i = 1, 3 do
        print("i:", i)
    end

    -- Generic for
    for key, value in pairs(Status) do
        print(key, value)
    end

    -- While loop
    local counter = 0
    while counter < 2 do
        counter = counter + 1
    end

    -- Repeat until
    repeat
        counter = counter - 1
    until counter == 0

    -- Goto + label
    local x = 0
    ::retry::
    x = x + 1
    if x < 2 then
        goto retry
    end

    -- Protected call
    local ok, err = pcall(function()
        error("Something went wrong")
    end)

    if not ok then
        print("Caught error:", err)
    end

    -- Local function inside scope
    local function inner()
        return "inner value"
    end

    print(inner())

    -- Anonymous function
    local squared = apply(function(n)
        return n * n
    end, 5)

    print("Squared:", squared)

    -- Multi-line string
    local text = [[
This is a multi-line string.
It spans multiple lines.
]]

    print(text)
end

main()
